server/src/main/java/org/elasticsearch/index/mapper/NestedValueFetcher.java

```diff
@@ -49,7 +50,14 @@ public NestedValueFetcher(String nestedField, FieldFetcher nestedFieldFetcher) {
         for (Object entry : nestedValues) {
             // add this one entry only to the stub and use this as source lookup
             stub.put(nestedFieldName, entry);
-            SourceLookup nestedSourceLookup = new SourceLookup();
+            SourceLookup nestedSourceLookup = new SourceLookup() {
+                @Override
+                public int docId() {
+                    // TODO instead of the original docId from the main doc, we probably need the ones for nested docs here
```

Comment:
I was able to remove this override and the test still passes. Maybe it's necessary in a slightly different set-up?

In any case, I understand wanting to avoid passing a nonsensical doc ID here. What would you think of using the typical method to initialize `SourceLookup`, which feels more robust:

```
SourceLookup nestedSourceLookup = new SourceLookup();
nestedSourceLookup.setSegmentAndDocument(context, lookup.docId());
nestedSourceLookup.setSource(filteredSource);
```

It would require hanging onto the `LeafReaderContext` supplied in `setNextReader`, but this doesn't seem too bad at all.

Reply:
sure, added this

Reply:
I had trouble with the FieldFetcherTests following this approach since there we would initially need to call `setSegmentAndDocument` at least once on the field fetcher, but for that test its difficult because we don't have a real reader etc... there or can mock it. In the end, and because I also don't remember what tests the "unset" docId in the newly created SourceLookup failed (the current ones pass) I completely removed the override.